src/main/java/edu/kit/provideq/toolbox/format/cnf/dimacs/ExpressionToDimacsCNF.java

```diff
@@ -0,0 +1,83 @@
+package edu.kit.provideq.toolbox.format.cnf.dimacs;
+
+import com.bpodgursky.jbool_expressions.And;
+import com.bpodgursky.jbool_expressions.Expression;
+import com.bpodgursky.jbool_expressions.Not;
+import com.bpodgursky.jbool_expressions.Or;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+class ExpressionToDimacsCNF {
+    private int nextVariable = 1;
+    private final HashMap<String, Variable> parsedVariables = new HashMap<>();
+
+    public DimacsCNF parse(Expression<String> cnfExpression) {
+        parsedVariables.clear();
+        nextVariable = 1;
+
+        var clauses = new ArrayList<ArrayList<Variable>>();
+
+        if (cnfExpression.getExprType().equals(And.EXPR_TYPE)) {
+            for (Expression<String> child : cnfExpression.getChildren()) {
+                clauses.add(parseClause(child));
+            }
+        } else {
+            clauses.add(parseClause(cnfExpression));
+        }
+
+        var variables = parsedVariables.values().stream().toList();
+
+        return new DimacsCNF(clauses, variables);
+    }
+
+    ArrayList<Variable> parseClause(Expression<String> expression) {
+        var variables = new ArrayList<Variable>();
+        addVariables(variables, expression);
+
+        return variables;
+    }
+
+    void addVariables(List<Variable> variables, Expression<String> e) {
+        switch (e.getExprType()) {
+            case Or.EXPR_TYPE -> {
+                for (Expression<String> child : e.getChildren()) {
+                    addVariables(variables, child);
+                }
+            }
+            case Not.EXPR_TYPE -> {
+                var name = ((com.bpodgursky.jbool_expressions.Variable<String>) e.getChildren().get(0)).getValue();
+                variables.add(getNegatedVariable(name));
+            }
+            case com.bpodgursky.jbool_expressions.Variable.EXPR_TYPE -> {
+                var name = ((com.bpodgursky.jbool_expressions.Variable<String>) e).getValue();
+                variables.add(getVariable(name));
+            }
+        }
+    }
+
+    Variable getNegatedVariable(String name) {
+        Variable var = parsedVariables.get(name);
+        if (var == null) {
+            var = new NegatedVariable(Math.abs(nextVariable), name);
```

Comment:
Yeah that's a bug.

It's a bit strange with the current setup, as the negation state in the `variables` of the DimacsCNF class can be quite random as it's a result of the order in which state they first turn up, and doesn't have any meaning for this purpose.
It might be better to have an interface or something that omits the negation state, but I didn't add it yet because it messes with Java's record implementation again.